### ISSUE TITLE
Django CMSPlugin.position null raises unorderable types: NoneType() < int()

### ISSUE THREAD

**Issue by pawelmarkowski** [divio/django-cms#4323]
Traceback (python 33 dj 183 cms 3.1.2) latest releases of plugins from pip:

```
...
File "/srv/venv/py33dj183/lib/python3.3/site-packages/cms/templatetags/cms_tags.py" in render_tag
  302.             content = get_placeholder_content(context, request, page, name, inherit, nodelist)
File "/srv/venv/py33dj183/lib/python3.3/site-packages/cms/templatetags/cms_tags.py" in get_placeholder_content
  221.         placeholder = _get_placeholder(current_page, page, context, name)
File "/srv/venv/py33dj183/lib/python3.3/site-packages/cms/templatetags/cms_tags.py" in _get_placeholder
  200.         assign_plugins(context['request'], fetch_placeholders, page.get_template(),  get_language())
File "/srv/venv/py33dj183/lib/python3.3/site-packages/cms/utils/plugins.py" in assign_plugins
  80.         plugin_groups[group] = build_plugin_tree(plugin_groups[group])
File "/srv/venv/py33dj183/lib/python3.3/site-packages/cms/utils/plugins.py" in build_plugin_tree
  138.                       key=attrgetter('parent_id', 'position'))
 
Exception Type: TypeError at /
Exception Value: unorderable types: NoneType() < int()
```

1. Restore the database
2. Run the command `manage.py cms fix-tree`
3. Make alias of plugin instance and put into the placeholder
4. Click "publish"
and the problem occures each time in my case ... after step 3 and 4. Script `manage.py cms fix-tree` is solving that problem temporary
In webbrowser console I can see: 
a) Uncaught TypeError: Cannot read property 'placeholder_id' of undefined cms.structuredboard.js:330 which is id = item.data('settings').placeholder_id;
b) GET http://xxxx.xxx/?edit_off 500 (INTERNAL SERVER ERROR)
c) cms.toolbar.js:234 Uncaught Error: [object Object]

regards

**Comment by ojii** [comment]
For the record, IRC logs: https://botbot.me/freenode/django-cms/2015-07-28/?msg=45802968&page=3

**Comment by pawelmarkowski** [comment]
A problem is not in the copy of a plugin instance but in his alias. Click right context menu in a structure view and choose option "Alias". Put alias into a placeholder and try to click publish... This steps should reproduce the error.

**Comment by ojii** [comment]
My quick glance at this brought me to ``CMSPlugin.position``, which has ``null=True``. The ``None`` ``position`` is what causes the error above.  Why is ``position`` allowed to be ``NULL``? I can't think of a reason why it should be ``NULL`` (or what that would mean within the CMS), so why don't we enforce valid values on a DB/validation level?

/cc @yakky

**Comment by pawelmarkowski** [comment]
I changed model CMSPlugin.position - now it can not be null and here is traceback

Traceback (most recent call last):
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/utils/decorators.py", line 110, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/views/decorators/cache.py", line 57, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/contrib/admin/sites.py", line 233, in inner
    return view(request, *args, **kwargs)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/utils/decorators.py", line 34, in _wrapper
    return bound_func(*args, **kwargs)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/views/decorators/http.py", line 45, in inner
    return func(request, *args, **kwargs)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/utils/decorators.py", line 30, in bound_func
    return func.__get__(self, type(self))(*args2, **kwargs2)
  File "/opt/rh/python33/root/usr/lib64/python3.3/contextlib.py", line 29, in inner
    return func(*args, **kwds)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/reversion/revisions.py", line 297, in do_revision_context
    return func(*args, **kwargs)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/cms/admin/pageadmin.py", line 959, in publish_page
    published = page.publish(language)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/cms/models/pagemodel.py", line 609, in publish
    self._copy_contents(public_page, language)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/cms/models/pagemodel.py", line 309, in _copy_contents
    copy_plugins_to(plugins, ph, no_signals=True)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/cms/utils/copy_plugins.py", line 18, in copy_plugins_to
    for old in old_plugins]
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/cms/utils/copy_plugins.py", line 18, in <listcomp>
    for old in old_plugins]
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/cms/models/pluginmodel.py", line 311, in copy_plugin
    new_plugin.save()
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/cms/models/pluginmodel.py", line 232, in save
    self.parent.add_child(instance=self)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/treebeard/mp_tree.py", line 970, in add_child
    return MP_AddChildHandler(self, **kwargs).process()
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/treebeard/mp_tree.py", line 361, in process
    newobj.save()
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/cms/models/pluginmodel.py", line 240, in save
    super(CMSPlugin, self).save()
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/db/models/base.py", line 710, in save
    force_update=force_update, update_fields=update_fields)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/db/models/base.py", line 738, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/db/models/base.py", line 822, in _save_table
    result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/db/models/base.py", line 861, in _do_insert
    using=using, raw=raw)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/db/models/manager.py", line 127, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/db/models/query.py", line 920, in _insert
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/db/models/sql/compiler.py", line 974, in execute_sql
    cursor.execute(sql, params)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/db/backends/mysql/base.py", line 129, in execute
    six.reraise(utils.IntegrityError, utils.IntegrityError(*tuple(e.args)), sys.exc_info()[2])
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/utils/six.py", line 658, in reraise
    raise value.with_traceback(tb)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/django/db/backends/mysql/base.py", line 124, in execute
    return self.cursor.execute(query, args)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/MySQLdb/cursors.py", line 220, in execute
    self.errorhandler(self, exc, value)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorvalue
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/MySQLdb/cursors.py", line 209, in execute
    r = self._query(query)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/MySQLdb/cursors.py", line 371, in _query
    rowcount = self._do_query(q)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/MySQLdb/cursors.py", line 335, in _do_query
    db.query(q)
  File "/srv/venv/py33dj183/lib/python3.3/site-packages/MySQLdb/connections.py", line 280, in query
    _mysql.connection.query(self, query)
django.db.utils.IntegrityError: (1048, "Column 'position' cannot be null")

**Comment by yakky** [comment]
@ojii ``null=True`` dates back to the very early implementation of the CMS. Being dependent on mptt/treebeard values, and given the current way ``CMSPlugin``s are created, it **must** be null.
@pawelmarkowski could this be related to #4318?

**Comment by ojii** [comment]
I know, the ``null=True`` was added back early 2009. Can't treebeard insert unsaved objects into a tree at a position? Does it need to be pre-saved?

**Comment by yakky** [comment]
After the sane add branch is merged we might consider this as we would have all the information from the very beginning and we can initialise the plugin properly, but I think it's not the root of @pawelmarkowski issue

**Comment by ojii** [comment]
@yakky sounds good, once #3706 is merged I can do some git magic and update #3804 with working tests and all those nice things.

**Comment by pawelmarkowski** [comment]
@yakky it is not the same problem. CMSPlugin position is Null in database after publishing my draft in 3.1.2

I added plugin_instance.position = new_plugin.position in CMSPlugin.copy_plugin method after line plugin_instance.depth = new_plugin.depth and it solved the problem but I am not sure is it correct way to fix that.

**Comment by pawelmarkowski** [comment]
above solution was wrong because position has to be recalculated ... but somewhere between CMSPlugin.copy_plugin (in plugin_instance) and save method error exists... I will prepare pull request with not null CMSPlugin.position in CMS and we will see which tests will fail, and where are we trying to put null value of position.

**Comment by pawelmarkowski** [comment]
CMSPlugin.position not null + default 0 passes tests and solving my issue.

**Comment by ojii** [comment]
Thanks for all the work/investigating you've done so far! What would be really useful would be a test that shows your problem (that ideally fails on current develop branch), so we can ensure it doesn't happen again in the future.

**Comment by yakky** [comment]
@pawelmarkowski thanks for this. We definitely need a test here: there should not be any case in which the position is not set, so another issue may be behind the first error.

**Comment by pawelmarkowski** [comment]
I am on holiday till 9th of August ... I will prepare a test or catch you at IRC

**Comment by trawick** [comment]
I have django-cms==3.1.4; somehow a customer created a plugin which appears as ``Text <Empty>`` and has position=None, after which rendering the page generated this error.  I guess the fix to disallow saving a plugin with position=None will eventually be in the 3.1.x branch???  For now, anyone with the issue that can't/doesn't want to figure out which plugin to delete from the failing page can try this:

```
diff --git a/cms/utils/plugins.py b/cms/utils/plugins.py
index fcb3cf3..a7498ba 100644
--- a/cms/utils/plugins.py
+++ b/cms/utils/plugins.py
@@ -134,7 +134,8 @@ def build_plugin_tree(plugins):
     """
     cache = dict((p.pk, p) for p in plugins)
     by_parent_id = attrgetter('parent_id')
-    nonroots = sorted(filter(by_parent_id, cache.values()),
+    nonroots = sorted(filter(lambda x: x.position is not None,
+                             filter(by_parent_id, cache.values())),
                       key=attrgetter('parent_id', 'position'))
     families = ((cache[parent_id], tuple(children))
                 for parent_id, children
```

I don't know specifically what "Text <Empty>" means; here's how it appears in the structure editor:

![image](https://cloud.githubusercontent.com/assets/1207350/12036322/d6241372-ae12-11e5-8ab0-563729a03c6c.png)

For this particular plugin, position was None.

**Comment by pawelmarkowski** [comment]
Hello,
@trawick djangocms_text_ckeditor generates this problem (with Null position). Example from sql log:

INSERT INTO `cms_cmsplugin` (`path`, `depth`, `numchild`, `placeholder_id`, `parent_id`, `position`, `language`, `plugin_type`, `creation_date`, `changed_date`) VALUES ('07IL0001', 2, 0, 3061, 32748, NULL, 'en', 'TextPlugin', '2016-01-18 14:51:07', '2016-01-18 14:51:07')

I have to take a look on that plugin. But right now problem is in CMS - I can not publish changes on fresh install.


--------------------------------
why position might be null here:  
...cms/models/pluginmodel.py
position = models.PositiveSmallIntegerField(_("position"), blank=True, null=True, editable=False)

**Comment by pawelmarkowski** [comment]
position = models.PositiveSmallIntegerField(_("position"), default = 0, editable=False) # solves my problem, but we have to investigate source

**Comment by jonasrogert** [comment]
I see the exact same issue.

Running fix-tree command unpublishes all pages and the site works. After publishing the home page the site will throw a 500 error unorderable types: NoneType() < int().

After i published i can see a row with position null in the cms_cmsplugin table (on a FilerImagePlugin).

**Comment by jribbens** [comment]
I was having this exact same thing recently. I've no idea how the null got in there but it meant the CMS couldn't display my page. In the end I had to manually poke a 0 into the database.

It seems to me that the code really needs fixing so it doesn't crash on null, rather than trying to find every one of apparently many places where nulls might arise and fix them all.

**Comment by mkoistinen** [comment]
We've made significant improvements in 3.3 that should affect these conditions. In particular, "ghost" plugins shouldn't happen anymore (under normal conditions) and in general there's been a lot of attention towards preventing and detecting tree corruption.

Can anyone affected by this in previous versions, please use `./manage.py cms fix-tree` (or the equivalent in their CMS version) and upgrade to version 3.3.

Closing this for now, please reopen or create a new ticket if you experience this issue under 3.3.0 (or the tip of the 'develop' branch). Many of these were fixed in 'release/3.2.x' as well if you're not ready to drop support for Django 1.6 and 1.7.

**Comment by mikemcgowan** [comment]
I've seen this issue just now on Django 1.8.13, Django-CMS 3.2.5, MySQL.

After publishing the home page and then selecting "Disable toolbar" from the CMS toolbar, I get:

```
File "/home/vagrant/.virtualenvs/globalmba_py3/lib/python3.4/site-packages/cms/templatetags/cms_tags.py" in _get_placeholder
  206.         assign_plugins(context['request'], fetch_placeholders, page.get_template(),  get_language())
File "/home/vagrant/.virtualenvs/globalmba_py3/lib/python3.4/site-packages/cms/utils/plugins.py" in assign_plugins
  80.         plugin_groups[group] = build_plugin_tree(plugin_groups[group])
File "/home/vagrant/.virtualenvs/globalmba_py3/lib/python3.4/site-packages/cms/utils/plugins.py" in build_plugin_tree
  138.                       key=attrgetter('parent_id', 'position'))

Exception Type: TypeError at /en/
Exception Value: unorderable types: NoneType() < int()
```

This doesn't happen if the toolbar is activated/present.
It only happens when it's not.

**Comment by pawelmarkowski** [comment]
Why won't we append default value in cms models (as I wrote in previous posts) for this exception? In my opinion it occurs due to leaks in plugins, which can not ensure default or proper position value. Otherwise django cms might be broken what leads to issues like this one (probably plugin is stable in 2.7, but for 3.4 version not).

I believe that is better to have a problem with unproper plugin position than broken CMS :)

**Comment by pawelmarkowski** [comment]
I have the same problem like mikemcgowan. I made project update to dj 196 and CMS 3.2.5 and  I received an error message while I was trying to publish old page. 

File "...cms/models/pluginmodel.py", line 248, in save
    super(CMSPlugin, self).save(*args, **kwargs)


File "..../cms/models/pluginmodel.py", line 248, in save
    super(CMSPlugin, self).save(*args, **kwargs)
